component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java

```diff
@@ -0,0 +1,222 @@
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.PREVENT_TOKEN_REUSE;
+
+/**
+ * Client Authentication handler to implement oidc private_key_jwt client authentication spec
+ * http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication.
+ */
+public class PrivateKeyJWTClientAuthenticator extends AbstractOAuthClientAuthenticator {
+
+    private static final Log log = LogFactory.getLog(PrivateKeyJWTClientAuthenticator.class);
+
+    JWTValidator jwtValidator;
+
+    public PrivateKeyJWTClientAuthenticator() {
+
+        jwtValidator = createJWTValidator();
+    }
+
+    /**
+     * To check whether the authentication is successful.
+     *
+     * @param httpServletRequest      http servelet request
+     * @param content                 map of request body params
+     * @param oAuthClientAuthnContext oAuthClientAuthnContext
+     * @return true if the authentication is successful.
+     * @throws OAuthClientAuthnException
+     */
+    @Override
+    public boolean authenticateClient(HttpServletRequest httpServletRequest, Map<String, List> content,
+                                      OAuthClientAuthnContext oAuthClientAuthnContext) throws OAuthClientAuthnException {
+
+        String clientAssertion = null;
+        List oauthJWTAssertion = content.get(OAUTH_JWT_ASSERTION);
+        if (oauthJWTAssertion != null && oauthJWTAssertion.get(0) != null) {
+            clientAssertion = oauthJWTAssertion.get(0).toString();
+        }
+        try {
+            boolean isValidAssertion = jwtValidator.isValidAssertion(getSignedJWT(clientAssertion));
+            if (StringUtils.isEmpty(oAuthClientAuthnContext.getClientId())) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Couldn't find client Id in context. Hence returning false.");
+                }
+                return false;
+            }
+            if (!isValidAssertion) {
+                return false;
+            }
+            return true;
+
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException(e.getMessage(), OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+    }
+
+    /**
+     * Returns whether the incoming request can be handled by the particular authenticator.
+     *
+     * @param httpServletRequest      http servelet request
+     * @param content                 map of request body params
+     * @param oAuthClientAuthnContext oAuthClientAuthnContext
+     * @return true if the incoming request can be handled.
+     */
+    @Override
+    public boolean canAuthenticate(HttpServletRequest httpServletRequest, Map<String, List> content,
+                                   OAuthClientAuthnContext oAuthClientAuthnContext) {
+
+        String clientAssertionType = null;
+        String clientAssertion = null;
+        List oauthJWTAssertionType = content.get(OAUTH_JWT_ASSERTION_TYPE);
+        List oauthJWTAssertion = content.get(OAUTH_JWT_ASSERTION_TYPE);
```

Comment:
I think we have a utility to get string empty of these parameters. Shall we use it ? You don't need to manually do it. It's there in abstract authenticator